docs/architecture/model_formats.md

```diff
@@ -0,0 +1,690 @@
+- `save_pretrained()`: Saves a speculator model to a local directory with optional uploading to the Hugging Face Hub
+- `push_to_hub()`: Directly pushes a model to the Hugging Face Hub for sharing with the community
+
+For detailed usage and examples, refer to the [entrypoints documentation](./entrypoints.md).
+
+#### SpeculatorConfig Class
+
+The `SpeculatorConfig` class extends the Transformers `PretrainedConfig` class, providing compatible APIs with the following methods:
+
+- `from_pretrained()`: Loads a speculator configuration from a local directory or the Hugging Face Hub
+- `save_pretrained()`: Saves a speculator configuration to a local directory
+- `push_to_hub()`: Directly pushes a configuration to the Hugging Face Hub for sharing with the community
+
+For detailed usage and examples, refer to the [entrypoints documentation](./entrypoints.md).
+
+## Examples
+
+This section provides examples of `config.json` files for popular speculative decoding algorithms.
+
+### Eagle
+
+```json
+{
+    "architectures": ["TransformerDraftModel"],
+    "model_type": "transformer_draft_model",
+    "torch_dtype": "bfloat16",
+    "transformers_version": "X.X.X",
+    "speculators_version": "X.X.X",
+    "inputs": ["input_embeddings", "hidden_states[-2]"],
+    "inputs_hidden_states_after_layer_norm": false,
+    "transformer_architecture": "LlamaDecoderLayer",
+    "transformer_input_type": "projection",
+    "transformer_include_output_layer_norm": false,
+    "attention_bias": false,
+    "attention_dropout": 0.0,
+    "bos_token_id": 128000,
+    "eos_token_id": [
+        128001,
+        128008,
+        128009
+    ],
+    "head_dim": 128,
+    "hidden_act": "silu",
+    "hidden_size": 4096,
+    "intermediate_size": 14336,
+    "max_position_embeddings": 131072,
+    "mlp_bias": false,
+    "num_attention_heads": 32,
+    "num_key_value_heads": 8,
+    "rms_norm_eps": 1e-05,
+    "rope_scaling": {
+        "factor": 8.0,
+        "high_freq_factor": 4.0,
+        "low_freq_factor": 1.0,
+        "original_max_position_embeddings": 8192,
+        "rope_type": "llama3"
+    },
+    "rope_theta": 500000.0,
+    "use_cache": true,
+    "vocab_size": 128256,
+    "speculators_config": {
+        "algorithm": "eagle",
+        "proposal_methods": [
+            {
+                "proposal_type": "greedy",
+                "draft_tokens": 5
+            },
+            {
+                "proposal_type": "sample",
+                "draft_tokens": 5,
+                "temperature": 0.8,
+                "top_p": 0.5
+            },
+            {
+                "proposal_type": "tree",
+                "tree_type": "static",
+                "initial_branching_factor": 4,
+                "branching_factor": 2,
+                "depth": 5
+            }
+        ],
+        "default_proposal_method": "tree",
+        "verifier": {
+            "name_or_path": "meta-llama/Llama-3.1-8B-Instruct",
+            "architectures": ["LlamaForCausalLM"],
```

Comment:
Thinking through the implications of this ...

Is this supposed to override the verifier model config (assuming they don't match?)

What would the use case be to override?

I'd imagine it would be quite tricky to make this override work in vLLM - by the time we load this config, we've probably done some setup of the verifier based on its config

If it's not an override, does vLLM just ignore it? That would be a bit weird ...